### PR TITLE
Add a scrollbar to the disassembly window

### LINKED ISSUE #2907: Scroll bar inside disassembly window.
**Is your feature request related to a problem? Please describe.**
Problem: There is no scroll bar inside the disassembly window which does not give any information about the size of disassembled code.

**Describe the solution you'd like**
Implementation of functional scroll bar inside the disassembly window UI.

I attached some examples of the required scroll bar.
scroll_bar
scroll_bar_issue

### PR DESCRIPTION
**Pull request by Anohkka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

Implements a scroll bar for the disassembly window. Helps with viewing different sections of a file in multiple disassembly windows, and is useful for scrolling through the binary during a debug session.

**Test plan (required)**

Tested with multiple, different binaries:

- Verified that scroll bar stays synchronized with the visual nav bar while seeking through the visual nav bar.
- Verified that multiple disassembler windows have separate scroll bars.
  - Seeking on one disassembler window will move every disassembler window to the selected address. This is the current behavior and has been left as is.
- Verified that scrolling with the scroll bar and Page Up / Page Down are both unchanged - they do not snap to the scroll bar's range.
- Verified that the scroll bar's range correctly maps to the binary in memory while debugging.
- Verified that the scroll bar's range correctly maps to the binary while emulating it.

Things I could not test:

- Loading a file without bin information, since that option is broken on my system (see #3249).
- Remote debugging, also broken on my system, even though it works fine on Rizin. The way I implemented the scroll bar range *should* work fine for remote debugging as well though.

**Closing issues**

Closes #2907.

### CODE REVIEW

**Comment by karliss** [comment]
This needs a lot more serious testing and a test plan across various usecases supported by cutter/rizin as mentioned in the mattermost discussion.

**Review by karliss — CHANGES_REQUESTED** [review]
Just playing with it for 10minutes made it break in so many different ways there is no point in me listing them or doing further review until you come up with more serious test plan and go through with it fixing any problems you find during the process.

Try actually using the Cutter (with the new scrollbar) and not just opening single unspecified file and doing 5 seconds of scrolling without paying attention what you are looking at beyond the fact that it changes somehow as you scroll. Try actually reading the disassembly while using the scrollbar. If you have no reverse engineering experience or understanding about various executable formats,  feel free to  spend a week trying to learn basics first and then return to this PR, otherwise it will be very hard for you to evaluate whether stuff you see makes any sense.

**Comment by wargio** [comment]
it's ok karliss. it's not a simple issue. they are trying to fix some issues to get used to the code.

**Comment by Anohkka** [comment]
Karliss, first of all, thank you for taking the time to test my PR.

I'm aware of my limitations when it comes to testing PRs here - I'm not a reverse engineer, and even a week using Cutter would not be enough to find every edge case. I had assumed that tying the scroll bar to the disassembly window refresh function would be enough, but it clearly was not.

It would be very helpful if you listed all the ways you made the scroll bar break, but I understand your frustration. I will spend more time with Cutter, prodding around for bugs in the scroll bar, in order to iron out the remaining kinks in its implementation.

File: src/widgets/DisassemblyWidget.cpp [hunk 1]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
-DisassemblyScrollArea::DisassemblyScrollArea(QWidget *parent) : QAbstractScrollArea(parent) {}
+DisassemblyScrollArea::DisassemblyScrollArea(QWidget *parent) : QAbstractScrollArea(parent)
+{
+    beginOffset = RVA_INVALID;
+    endOffset = RVA_INVALID;
+    fetchStats();
+    connect(Core(), &CutterCore::refreshAll, this, &DisassemblyScrollArea::fetchStats);
+    connect(Core(), &CutterCore::functionsChanged, this, &DisassemblyScrollArea::fetchStats);
+    connect(Core(), &CutterCore::flagsChanged, this, &DisassemblyScrollArea::fetchStats);
+    connect(Core(), &CutterCore::globalVarsChanged, this, &DisassemblyScrollArea::fetchStats);
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Why update after function/flag/global var changes? I don't see how any of those could affect scrollable range.

File: src/widgets/DisassemblyWidget.cpp [hunk 2]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
+        RzPVector *mapsPtr = rz_io_maps(core->io);
+        if (!mapsPtr) {
+            setVerticalScrollBarPolicy(Qt::ScrollBarPolicy::ScrollBarAlwaysOff);
+            return;
+        }
+        CutterPVector<RzIOMap> maps { mapsPtr };
+        for (const RzIOMap *const map : maps) {
+            if (Core()->currentlyEmulating && std::strncmp(rz_str_get(map->name), "mem.", 4) == 0) {
+                continue;
+            }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp:864]: Might be nice to add a comment why "mem." is getting skipped. Is that some kind of emulated heap or stack region?

File: src/widgets/DisassemblyWidget.cpp [hunk 3]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
+{
+    const QSignalBlocker blocker(verticalScrollBar());
+    if (RVA stepSize = getVStepSize()) {
+        verticalScrollBar()->setValue((address - beginOffset) / stepSize);
+    } else {
+        setVerticalScrollBarPolicy(Qt::ScrollBarPolicy::ScrollBarAlwaysOff);
+    }
+}
+
+void DisassemblyScrollArea::fetchStats()
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: This could be named a bit better. Maybe something like refreshScrollbar or fetchScrollbarStats.

File: src/widgets/DisassemblyWidget.cpp [hunk 4]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
+                endOffset = e;
+            }
+        }
     }
+    if (endOffset == 0) {
+        beginOffset = 0;
+    }
+    verticalScrollBar()->setMinimum(0);
+    if ((endOffset - beginOffset) > 100) {
+        verticalScrollBar()->setMaximum(100);
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Why artificially limit the scroll range to 100 steps? It makes scrollbar to jump in much bigger steps than it should.

There needs to be some limit as QScrollbar doesn't support 64bit values, but it shouldn't be 100.

File: src/widgets/DisassemblyWidget.cpp [hunk 5]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
+    } else {
+        verticalScrollBar()->setMaximum(endOffset - beginOffset);
+    }
+    if (getVStepSize() <= 0) {
+        setVerticalScrollBarPolicy(Qt::ScrollBarPolicy::ScrollBarAlwaysOff);
+        return;
+    }
+    setVerticalScrollBarPolicy(Qt::ScrollBarPolicy::ScrollBarAsNeeded);
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Does `ScrollBarAsNeeded` even make sense in combination with scrollbar being reused for memory range? Wouldn't it result in making decisions based on rendered pixel content height which is completely separate thing from the byte logic which for which this PR uses it?

File: src/widgets/DisassemblyWidget.cpp [hunk 6]
```diff
@@ -736,29 +739,115 @@ void DisassemblyWidget::setupColors()
...
-    verticalScrollBar()->setRange(0, 10);
-    verticalScrollBar()->setValue(5);
-    verticalScrollBar()->blockSignals(false);
+    if (event->angleDelta().y() > 0) {
+        emit scrollLines(-5);
+        return;
+    } else if (event->angleDelta().y() < 0) {
+        emit scrollLines(5);
+        return;
+    }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Why 5? Also this is ignoring the actual scroll amount reported.

The typical scroll speed is 3lines per single 15(deg) mouse tick. So it would have to be something like.

```
if (event->angleDelta().y != 0) {
    emit scrollLines(event->angleDelta().y / 40); // 3 lines for 15 degres, 1 line for 5 degrees, angle reported in 1/8 of degree . 8*5=40
}
```

But since high resolution mice are thing, need to have a bit more complicated logic accumulating angleDelta until +-40 units are reached.

**Review by karliss — CHANGES_REQUESTED** [review]
Looking more promising already.

Compared to initiatial version doesn't crash when opening small files anymore, works better when debugging, doesn't use completely nonsensical range when opening PE(.exe) executables. 


There seems to be some bug with handling end of range. Here are the steps I did:
1) copy and paste 0x00008a0b copies of "90" in open shellcode tab and open it, exact count probably doesn't matter but it should be on the same order of magnitude
2) reduce the height of disassembly window as much as possible
3) attempt scrolling to the bottom using scrollbar

Expected result:
Instruction at 0x00008a0b is either somewhere in screen or one instruction above the first visible instruction. (Not asking for it to be perfectly at bottom, as that's currently not easily possible due to line!=byte) .

Actual result:
last visible instruction is at 0x89ef, cutting off some of the last instructions.

image
image


Another problem I observed is where you can't reach use scrollbar to reach the top under certain conditions. Something like:

1) Jump to the middle of executable
2) jump to an address 20 bytes after the scrollbar top using "Type flag name or address here".
3) try to drag the scrollbar up to reach the actual top -> position doesn't change


I am guessing that both problems are actually caused by bad edge case handling in with the max 100 and `currentVScrollAddr()` logic. Once you bump the max value as requested in other comment, it might look like the problem disasppeared. But it will likely still be there, just harder to trigger. So I recommend fixing these 2 first and only then dealing with raising max scroll value.

File: src/widgets/DisassemblyWidget.cpp (hunk 1 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: I took those connections from VisualNavbar. Since they don't affect scrollable range, I'll remove them.

**Comment by Anohkka** [comment]
Both issues should be fixed. Let me know if there's anything else that needs work.

File: src/widgets/DisassemblyWidget.cpp [hunk 7]
```diff
@@ -736,29 +740,164 @@ void DisassemblyWidget::setupColors()
...
-    verticalScrollBar()->setRange(0, 10);
-    verticalScrollBar()->setValue(5);
-    verticalScrollBar()->blockSignals(false);
+    // Handle scroll direction changes
+    if (accumScrollWheelDeltaY > 0 && event->angleDelta().y() < 0) {
+        accumScrollWheelDeltaY = 0;
+    }
+    if (accumScrollWheelDeltaY < 0 && event->angleDelta().y() > 0) {
+        accumScrollWheelDeltaY = 0;
+    }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Is this really necessary to complicate things with this. Shouldn't it be fine to let the negative numbers cancel out positive ones by simply adding up?

File: src/widgets/DisassemblyWidget.cpp [hunk 8]
```diff
@@ -736,29 +740,164 @@ void DisassemblyWidget::setupColors()
...
+    accumScrollWheelDeltaY += event->angleDelta().y();
+    int lineCount = 0;
+    if (accumScrollWheelDeltaY >= 40 || accumScrollWheelDeltaY <= -40) {
+        lineCount = accumScrollWheelDeltaY / 40;
+        accumScrollWheelDeltaY -= 40 * lineCount;
+    }
+    if (lineCount) {
+        emit scrollLines(-lineCount);
+    }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]:
```suggestion
    if (accumScrollWheelDeltaY >= 40 || accumScrollWheelDeltaY <= -40) {
        int lineCount = accumScrollWheelDeltaY / 40;
        accumScrollWheelDeltaY -= 40 * lineCount;
        emit scrollLines(-lineCount);
    }
```

File: src/widgets/DisassemblyWidget.cpp [hunk 9]
```diff
@@ -736,29 +740,164 @@ void DisassemblyWidget::setupColors()
...
+            if (e > endOffset) {
+                endOffset = e;
+            }
+        }
     }
+    if (endOffset == 0) {
+        beginOffset = 0;
+    }
+    verticalScrollBar()->setMinimum(0);
+    if ((endOffset - beginOffset) > 100000) {
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Magic numbers, same for 40.  Would be good to define constants (doing it locally is fine) and explain how they were chosen.

**Review by karliss — CHANGES_REQUESTED** [review]
Don't forget to fix CI failures.

There is something very weird happening near the end. 

Open shellcode with 0x00000168 bytes an slowly drag the scrollbar to bottom. Near the end it suddenly jumps to the end, but keeps reacting mouse movement. 

Probably related, if you slowly scroll with a scrollwheel of such file (tested with 0x87), there is point near end where scrolling by 3 lines suddenly cause scrollbar position to jump by a lot. I am guessing some kind of inconsistency in scrolbar->address, adress->scrollbar mapping in combination with callback loop (scrollbar updates seek position, seek position updates scrollbar pos).

~Problem is even more exaggerated if you open raw binary file (64 bytes) and in advanced options during opening set map offset to 0x1000~ (it's equally bad for the same file without setting map offset)

With small files for example 43, there is a different 2 byte jump near the end. As you scroll with scrollbar, the top position changes 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2b.  , skipping 0x2a. Not sure if it's related or separate problem from the one described above.

File: src/widgets/DisassemblyWidget.cpp (hunk 7 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: It makes scrolling more consistent. I found that without it, sometimes scrolling down once then scrolling back up once (or the other way around) would sometimes scroll to a different line, instead of returning to the line you started on.

Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: Ok, I'll play with it a bit to see how it feels for me. In the mean time while doing rest of fixes at least merge them into single if statement.

Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: > I found that without it, sometimes scrolling down once then scrolling back up once (or the other way around) would sometimes scroll to a different line, instead of returning to the line you started on.

Probably because you are not initializing  `accumScrollWheelDetalY`.  Having an offset, results in getting 2 instead of 3 lines when crossing 0, mix of movement by 2 and 3 might explain what you observed.

The code above resetting accumulator to 0, removes  the offset you get from it not being properly initialized.

**Comment by Anohkka** [comment]
Addressed those issues as well. Scrolling to the bottom now should place the binary's end offset at the top of the disassembly viewer. Not ideal, but it's hard to do better. The only other option I see is to perform a binary search using CutterCore::disassembleLines with different offsets until we find an offset where the binary's end offset is at the last line.

**Comment by karliss** [comment]
> Addressed those issues as well. Scrolling to the bottom now should place the binary's end offset at the top of the disassembly viewer. Not ideal, but it's hard to do better.

Like i said in https://github.com/rizinorg/cutter/pull/3430#pullrequestreview-2744979154 that's acceptable. I have some plans for improving line handling in disassembly widget but that's a separate giant task.

File: src/widgets/DisassemblyWidget.cpp [hunk 10]
```diff
@@ -736,29 +740,152 @@ void DisassemblyWidget::setupColors()
...
+        --endOffset;
+    }
+    if (endOffset == 0) {
+        beginOffset = 0;
+    }
+    verticalScrollBar()->setMinimum(0);
+    // Maximum value recommended by Qt
+    // see https://doc.qt.io/qt-6/qscrollbar.html
+    // The greater this value, the smaller a file must be for the scroll bar to stay accurate
+    // A rangeMax of 100000 lets the scroll bar handle files up to ~167.8TB in size without issue
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: Read the Qt docs and think why it suggest 100000.  Note that it says 100000 pixels not counts and that scrollbar itself has no issues working with larger numbers within limits of int.

I am 70% sure that logic behind that suggestion is something like this:

Typical computer screen is currently 1000-2000 px high. Mouse coordinates are in screen pixels. If the scrollbar corresponded to 1000'000 pixels of screen content then dragging scrollbar up/down by single pixel would cause  content to move by  `1000000/1000=1000px` or by whole screen. At that point scrollbar isn't anymore useful for smoothly scrolling the content down as you read it. For 100000px it would cause movement by 100px or 1/10 of the screen height which while not smooth would at least allow keeping track of line you are reading.

Since our scrollbar manipulates bytes(approximating lines) not pixels we are already past the smooth scrolling limit by a lot.
Try opening a few thousand line source file in a code editor and navigating it using scrollbar. In my tests  ~3000lines is the limit where scrollbar can still be somewhat reasonably used for navigation, at 5000 it becomes quite uncomfortable and at 10000 unusable.  But the size of executable are what they are, we can't do much about it. There is still a bit of value from scrollbar seeing rough current position and rough scrubbing through the file to get the general feeling of executable.

By 167.8TB I assume you are referring to the code `currentVScrollAddr` where you have to switch to different formula to avoid multiplication overflow. 

Having the max value of 100K  has direct negative effect for files in 100KB-32MB range (which are a lot more likely than 160TB). I would even dare to claim that >100KB executable are quite common, especially if they are statically linking libraries. While single pixel of scrollbar movement already results in a large jump, you can still get single count change by clicking on arrows at the top/bottom (see the other comment about arrows). Having maxvalue less than binary size means that the jumps of single click are larger than what they would be for files under the limit, same for pageStep jumps.

So I suggest setting it as high as Qt scrollbar itself can handle. Let's say 512x1024x1024. That should cover typical executable size while not risking being too close to INT_MAX. I saw some multiplications by 2 in the QScrollbar code.

Also any issues for executables larger than (2^64)/scroll_max should be avoidable with little bit of math. 
Using `position * (range/scroll_max) + min(position, range % scroll_max)` should fix the end not being reachable. Not as evenly distributed as `(position * range)/scroll_max` but more than good enough for rarer cases like that.

If you need to test the behavior instead of generating giant dummy file, you can map multiple regions using `om` comment.

File: src/widgets/DisassemblyWidget.cpp [hunk 11]
```diff
@@ -50,7 +51,9 @@ DisassemblyWidget::DisassemblyWidget(MainWindow *main)
     layout->setContentsMargins(0, 0, 0, 0);
     mDisasScrollArea->viewport()->setLayout(layout);
     splitter->addWidget(mDisasScrollArea);
-    mDisasScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarPolicy::ScrollBarAlwaysOff);
+    QScrollBar *vScrollBar = mDisasScrollArea->verticalScrollBar();
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: pageStep can probably be raised to 40 from default 10.

I am wondering if scrollbar singleStep should be raised to 4.

On arm and many other fixed size instruction sets instructions are 4 maybe 2 bytes. Even on x86_64 many instructions are >1 byte. If you are using the scrollbar arrows to slightly advance by few lines while reading it, having to click 4 times just to advance by single instruction would be frustrating. Although unless you are using a mouse without scrollwheel and for some reason don't want to use keyboard, there are alternatives.

On the other hand if you are dealing x86_64 and want to precisely align instruction parsing to specific position,  advancing by single byte can be beneficial. And the scrollbar buttons would be one of the rare quick ways of doing it. Only other method for doing it, that I can think of is manually typing address in the address bar.

**Review by karliss — CHANGES_REQUESTED** [review]
Moving down using little arrows at bottom and top of scrollbar doesn't work in a ~200KB file. Moving up works fine.

Make sure to have chosen system qt theme which doesn't hide them, for example Fusion and Windows.

image

File: src/widgets/DisassemblyWidget.cpp (hunk 11 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: I opted for making it so the scroll bar buttons scroll through instructions. Currently 1 instruction per click, but the speed can easily be changed. Let me know if it's an improvement.

File: src/widgets/DisassemblyWidget.cpp (hunk 10 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: > Also any issues for executables larger than (2^64)/scroll_max should be avoidable with little bit of math.
Using position * (range/scroll_max) + min(position, range % scroll_max) should fix the end not being reachable. Not as evenly distributed as (position * range)/scroll_max but more than good enough for rarer cases like that.

This works fine for getting an address from a scroll bar position, but unless there's something I'm missing, solving for `position` (which we need to do when setting the scroll bar position from an address) still requires a dangerous multiplication.

File: src/widgets/DisassemblyWidget.cpp [hunk 12]
```diff
@@ -736,29 +739,187 @@ void DisassemblyWidget::setupColors()
...
+RVA DisassemblyScrollArea::currentVScrollAddr()
+{
+    int maximum = verticalScrollBar()->maximum();
+    if (!maximum || !binSize()) {
+        return beginOffset;
+    }
+    // Fallback formula for large files
+    if ((RVA_MAX / maximum) > binSize()) {
+        return verticalScrollBar()->value() * (binSize() / maximum)
+                + std::min(static_cast<RVA>(verticalScrollBar()->value()), binSize() % maximum)
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: You should be able to just explicitly specify template type to avoid ambiguity instead of doing a static cast.
`std::min<RVA>(a, b)`

File: src/widgets/DisassemblyWidget.cpp (hunk 10 above)
Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: No it doesn't you just need to do a bit of reordering, to achieve the same process as what happens in opposite direction.

File: src/widgets/DisassemblyWidget.cpp [hunk 13]
```diff
@@ -736,29 +739,187 @@ void DisassemblyWidget::setupColors()
...
+    verticalScrollBar()->setPageStep(40);
+    connect(verticalScrollBar(), &QScrollBar::actionTriggered, this, [this](int action) {
+        QScrollBar *vScrollBar = verticalScrollBar();
+        int val = vScrollBar->value();
+        // Check if val is vScrollBar->minimum() or vScrollBar->maximum()
+        // to prevent buttons from scrolling when the scroll bar when is either at
+        // the top or the botttom
+        if (val == vScrollBar->minimum() || val == vScrollBar->maximum()) {
+            return;
+        }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: >  // Check if val is vScrollBar->minimum() or vScrollBar->maximum()
// if (val == vScrollBar->minimum() || val == vScrollBar->maximum())`

Don't write comments which only repeat the code.

Also you are not really preventing buttons from scrolling, you are just skipping your extra code. The default behavior of in reaction to buttons will still happen resulting in scroll by singleStep.

Not scrolling down when you are at top doesn't make sense.

File: src/widgets/DisassemblyWidget.cpp [hunk 14]
```diff
@@ -736,29 +739,187 @@ void DisassemblyWidget::setupColors()
...
+        switch (action) {
+        case QAbstractSlider::SliderSingleStepAdd:
+            // Due to the way this signal works,
+            // setting the slider pos to its current value here
+            // prevents it from moving, allowing us to basically
+            // override the scroll bar buttons' behavior
+            // See https://doc.qt.io/qt-6/qabstractslider.html#actionTriggered
+            // for more info.
+            // Setting the scroll bar's single step to 0 prevents it from scrolling
+            // while hovering over it, so we have to use this workaround
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: I am guessing what you mean to say was that if you had set singleStep to 0, it would have broken a bunch of other stuff. So instead you are using this approach to override the default behavior. 
If it had worked, in my opinion setting singleStep to 0 to block the default behavior would have been as much of workaround as the stuff happening here.

Initially when I read it, I spent a few minutes searching where you are setting singleStep to 0. Interpreting the comment as "single step has been set to 0, so we must use this workaround to fix the negative sideffects caused by it".

It is not bad to mention alternative approaches you considered in PR discussion. But the way it's mentioned in this comment, just seems to cause more confusion. Might be be better to remove the last sentence.

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/widgets/DisassemblyWidget.cpp (hunk 10 above)
Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: Wasn't there there comment from you suggesting that `setVScrollPos` still requires a potentially overflowing multiplication?

File: src/widgets/DisassemblyWidget.cpp (hunk 11 above)
Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: Stepping by instructions like this is better is even better. Didn't expect there to be a way of easily overriding this part of QScrollbar behavior.

**Comment by Anohkka** [comment]
The scroll bar buttons work even if you're viewing something outside the scroll bar's range. They behave somewhat weirdly in that if you're past the end of the scroll bar's range and try to scroll up, it works fine, but scrolling down won't work. How should the scroll bar behave in such cases? Some options:

- Snapping back into the scroll bar address range if you press a scroll bar button

- Allowing the scroll bar buttons to scroll beyond the binary address range
  - Always? Only when debugging? Only when you're already viewing an address outside the scroll bar's range?

- Preventing use of the scroll bar buttons altogether while outside the scroll bar address range

- Keeping them as is

File: src/widgets/DisassemblyWidget.cpp (hunk 14 above)
Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: Please just remove that part.

**Comment by karliss** [comment]
I don't have any strong feeling with regards to button behavior outside the range, except that they shouldn't cause snapping behavior.

Working always (2.1) or as is towards range (4) would be fine with me.

File: src/widgets/DisassemblyWidget.cpp [hunk 15]
```diff
@@ -736,29 +739,183 @@ void DisassemblyWidget::setupColors()
...
+            // for more info.
+            // Setting the scroll bar's single step to 0 would allow us to
+            // override scroll bar button behavior too, but it prevents
+            // scrolling while hovering over the scroll bar, so it's not an option.
+            vScrollBar->setSliderPosition(val);
+            if (val != vScrollBar->maximum()) {
+                emit scrollLines(1);
+            }
+            return;
+        case QAbstractSlider::SliderSingleStepSub:
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp:764]: Looks like now you have an off by one mistake. When positioned at top it's possible to step single byte outside the range.

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/widgets/DisassemblyWidget.cpp (hunk 10 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: Yeah, I deleted it when I figured out how to get rid of the multiplication. Didn't want you to spend time on something I had already figured out.

File: src/widgets/DisassemblyWidget.cpp [hunk 16]
```diff
@@ -736,29 +739,185 @@ void DisassemblyWidget::setupColors()
...
+        return;
+    }
+    if ((RVA_MAX / maximum) < binSize()) {
+        // Fallback formula for large files
+        if (address < (binSize() % maximum + beginOffset)) {
+            scrollBarPos = 0;
+        } else {
+            scrollBarPos =
+                    ((address - beginOffset - (binSize() % maximum))) / (binSize() / maximum);
+        }
```
Comment (karliss) [inline src/widgets/DisassemblyWidget.cpp]: This is not the same mapping as one done by  `currentVScrollAddr()` does.  Less obvious due to being used only for `(RVA_MAX / maximum) < binSize()`, but if you add testing code to always use fallback path (for both) it becomes a lot more apparent how big the difference can be.

**Review by karliss — CHANGES_REQUESTED** [review]

File: src/widgets/DisassemblyWidget.cpp (hunk 16 above)
Comment (Anohkka) [reply src/widgets/DisassemblyWidget.cpp]: Interesting - I was actually considering making the fallback formula the only formula, since it appared to work so well in my tests. Seems like you won the edge case lottery. I wrote a small test program (willing to share its code if you'd like), and there's a noticeable deviation (I'm considering a single pixel of deviation in a 5k pixel tall screen to be "noticeable") in fewer than 0.01% of the test cases I've run - the binary size needs to be at a very specific range above the fallback formula threshold. These edge cases do exist though, and I've seen some where the deviation can be as large as ~1/3 of the entire scroll bar.

Clamping the scroll bar's range to a lower value *seems* to fix this - can't tell for sure since it's possible that it makes the edge cases just a lot less likely to appear - but we run into the same issue as before, where scrolling while hovering over the scroll bar scrolls with increased speed. I think that's fine, personally, since there's currently no way to scroll through a binary quickly (aside from the nav bar, which isn't of much use while debugging), and you'd still be able to scroll at a slower pace by hovering over the disassembly text itself. We might even be able to make the fallback formula the only formula, which would simplify the code.

Comment (karliss) [reply src/widgets/DisassemblyWidget.cpp]: From your description I am not quite sure what you are suggesting. 

> Seems like you won the edge case lottery

There is no lottery, it's called taking a piece of paper and using your head to analyze what the formula actually does,  instead of randomly making changes to code and relying on observations.

When faced with the question of how to place 999 stones in 500 boxes your choice was put one stone in each box and the remaining 499 in the first box.

> the binary size needs to be at a very specific range above the fallback formula threshold

If you  understand what's happening you will know that the problem happens almost always (except when max is a perfect divisor of range), it's just the amount of error that varies.
